src/abstractdataframe/join.jl

```diff
@@ -280,7 +286,22 @@ join(name, job2, on = :ID => :identifier)
 function Base.join(df1::AbstractDataFrame,
                    df2::AbstractDataFrame;
                    on::Union{<:OnType, AbstractVector{<:OnType}} = Symbol[],
-                   kind::Symbol = :inner, makeunique::Bool=false)
+                   kind::Symbol = :inner, makeunique::Bool=false,
+                   indicator::Union{Nothing, Symbol} = nothing)
+
+    if !(indicator === nothing)
+        names_in_use = Set([names(df1); names(df2); [indicator]])
+        indicator_cols = ["_left", "_right"]
+        for i in 1:2
+            while Symbol(indicator_cols[i]) ∈ names_in_use
+                 indicator_cols[i] = indicator_cols[i] * "X"
+            end
+         end
+        indicator_cols = Symbol.(indicator_cols)
+        df1 = hcat(df1, DataFrame(Dict(indicator_cols[1] => [true for i in 1:nrow(df1)] )))
```

Comment:
Wouldn't this be enough (and use less memory)?
```julia
df1[indicator_cols[1]] = trues(nrow(df1))
```
Also better call `Symbol` when indexing rather than replacing `indicator_cols` (type instability!).

Reply:
so just remembered why I was using `hcat` -- doing it the way proposed leads to modification of the input dataset. Fixing in #1421